### ISSUE TITLE
Wrong result when operating with tensors

### ISSUE THREAD

**Issue by donielix** [sympy/sympy#15676]
Hi everyone, and sorry for my bad english.

I reckon that there's a issue within the tensor module. 
I'm going to put the code that reproduces the wrong result to make it clear.
#-----------------------------Defining data------------------------------------
```
import sympy as sp
import sympy.tensor.tensor as tn
sp.init_printing()
g=tn.TensorIndexType('g')
g.data=[1,-1,-1,-1] #to define the Minkowski metric
i,j,k,l,m=tn.tensor_indices('i:m',g)
e=tn.tensorhead('epsilon',[g]*4,[[4]])
e.data=[[[[sp.LeviCivita(a,b,c,d) for d in range(4)] for c in range(4)] for b in range(4)] for a in range(4)] #to define the levi Civita tensor
```

#---------------------------------------------

```
>> e(i,j,k,l)[0,1,2,3]
1
>> e(-i,-j,-k,-l)[0,1,2,3]
1
```

In theory, I'd have to get -1 in the second entry because the Metric

I hope someone can fix it soon.
Thanks

**Comment by asmeurer** [comment]
You might take a look at the new tensor module that is in master. CC @Upabjojr

**Comment by Upabjojr** [comment]
How it looks on master:
https://github.com/sympy/sympy/blob/28eba540390e8f0afa4f1ee14c994cad6bd482e9/sympy/tensor/tensor.py#L1637

**Comment by Upabjojr** [comment]
The problem exists with the new API as well.

**Comment by Upabjojr** [comment]
Thanks for the report. I'm busy on other stuff right now, so we'll keep this issue open.

**Comment by Upabjojr** [comment]
> Can I work on this if you are busy @Upabjojr . :-)

OK!

**Comment by czgdp1807** [comment]
```
import sympy as sp
from sympy.tensor import tensor as tn
sp.init_printing()
g=tn.TensorIndexType('g')
i,j,k,l,m=tn.tensor_indices('i:m',g)
e=tn.tensorhead('epsilon',[g]*4,[[4]])
print(e(i, j, k, l).replace_with_arrays({e(i, j, k, l) : [[[[sp.LeviCivita(a,b,c,d) for d in range(4)] for c in range(4)] for b in range(4)] for a in range(4)], g : sp.diag(1, -1, -1, -1)}, [i , j, -k, l])[0, 1, 2, 3]) #to define the levi Civita tensor
```
The above code works correctly with 1.4
Though the issue still pertains.
Working on it, though quite lately. Apologies.

**Comment by czgdp1807** [comment]
```
expr = e(i, j, k, l) * M(-k, -y)
print(expr.replace_with_arrays({e(i, j, k, l) : LC, M(-k, -y) : [[1, 0, 0, 0], [0, -1, 0, 0], [0, 0, -1, 0], [0, 0, 0, -1]]}, [i, j, -y, l])[0, 1, 2, 3])
````
The above code give correct results for lowering `k`.
@Upabjojr I think there is a problem with `tensorproduct` or `tensorcontraction`.

**Comment by Upabjojr** [comment]
Can you find an explicit case where `tensorproduct` or `tensorcontraction` give the incorrect result?

**Comment by marc-git** [comment]
I have a problem with the product and contraction which I think is a real mistake (hopefully not my error) and I can give a concrete example. The fourth order identity tensor, when doubly contracted with a second order tensor should give the same second order tensor. By my way the fourth order identity is (how to write in latex?) delta_ik delta_jl e_i e_j e_k e_l . So I program it so:

```
ident  = MutableDenseNDimArray.zeros(3,  3, 3,3)
for i in range(0,3):
    for j in range(0,3):
        for k in range(0,3):
            for l in range(0,3):
                if i==k and j==l :
                    ident[i,j,k,l] = 1```

Looks like what i wanted. 

```
from sympy.abc import a,b,c,d,e,f,g,h,i
A = MutableDenseNDimArray([[a,b,c],[d,e,f],[g,h,i]])
tensorcontraction(tensorproduct(ident, A), (2,3,4,5))
```

This should reproduce A, but produces only the diagonal.  If I try the same with the fourth order diagonaliser produced as so:

```
P2 = MutableDenseNDimArray.zeros(3,  3, 3,3)
for i in range(0,3):
    P2[i,i,i,i] = 1
tensorcontraction(tensorproduct(P2, A), (2,3,4,5))
```
I get the same result, but how can this be correct when these are wildly different products?

If this is not the same problem I will file a new bug.

**Comment by Upabjojr** [comment]
> tensorcontraction(tensorproduct(ident, A), (2,3,4,5))

Indexing start from zero, not from 1.

```python
In [3]: from sympy.abc import a,b,c,d,e,f,g,h,i
   ...: A = MutableDenseNDimArray([[a,b,c],[d,e,f],[g,h,i]])
   ...: tensorcontraction(tensorproduct(ident, A), (1,2,3,4))
Out[3]:
⎡a  b  c⎤
⎢       ⎥
⎢d  e  f⎥
⎢       ⎥
⎣g  h  i⎦
```

**Comment by marc-git** [comment]
Then is it the case that we define A : B = A_ij B_ji ? I understand this usage is not uncommon, but the common usage in continuum theory is A:B = A_ij B_ij . Which forms a scalar product. I think the definition should be correctly described in the docs.

**Comment by Upabjojr** [comment]
> I understand this usage is not uncommon,

Yes, they are the traces `tr(A*B)` and `tr(A*B.T)`, respectively.

**Comment by marc-git** [comment]
> 
> Yes, they are the traces `tr(A*B)` and `tr(A*B.T)`, respectively.
Yes. I was trying to be generous. I never heard of the later interpretation and as far as I can tell it is unphysical. Either way the documentation should describe which of the two is being performed as they are not equivalent. Or at least not clearly so... https://math.stackexchange.com/questions/348739/double-dot-product-vs-double-inner-product

**Comment by simpla-fusion** [comment]
> Hi everyone, and sorry for my bad english.
> 
> I reckon that there's a issue within the tensor module. I'm going to put the code that reproduces the wrong result to make it clear. #-----------------------------Defining data------------------------------------
> 
> ```
> import sympy as sp
> import sympy.tensor.tensor as tn
> sp.init_printing()
> g=tn.TensorIndexType('g')
> g.data=[1,-1,-1,-1] #to define the Minkowski metric
> i,j,k,l,m=tn.tensor_indices('i:m',g)
> e=tn.tensorhead('epsilon',[g]*4,[[4]])
> e.data=[[[[sp.LeviCivita(a,b,c,d) for d in range(4)] for c in range(4)] for b in range(4)] for a in range(4)] #to define the levi Civita tensor
> ```
> 
> #---------------------------------------------
> 
> ```
> >> e(i,j,k,l)[0,1,2,3]
> 1
> >> e(-i,-j,-k,-l)[0,1,2,3]
> 1
> ```
> 
> In theory, I'd have to get -1 in the second entry because the Metric
> 
> I hope someone can fix it soon. Thanks
The same problem reappears in version 1.10.
